src/Workspaces/Core/Portable/Workspace/WorkspaceEventMap.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.Workspace;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed class WorkspaceEventMap
+{
+    private readonly SemaphoreSlim _guard = new(initialCount: 1);
+    private readonly Dictionary<WorkspaceEventType, EventHandlerSet> _eventTypeToHandlerSet = [];
+
+    public WorkspaceEventRegistration AddEventHandler(WorkspaceEventType eventType, WorkspaceEventHandlerAndOptions handlerAndOptions)
+    {
+        using (_guard.DisposableWait())
```

Comment:
fwiw, if you do not use this in an async or cancellable fashion, i'm not sure if this is any better than just a plain primitive lock.

but fine to keep if you prefer this.